src/function/maybe_changed_after.rs

```diff
@@ -1,12 +1,41 @@
 use crate::{
+    cycle::CycleRecoveryStrategy,
     key::DatabaseKeyIndex,
+    table::sync::ClaimResult,
     zalsa::{Zalsa, ZalsaDatabase},
     zalsa_local::{ActiveQueryGuard, EdgeKind, QueryOrigin},
     AsDynDatabase as _, Id, Revision,
 };
+use rustc_hash::FxHashSet;
 
 use super::{memo::Memo, Configuration, IngredientImpl};
 
+/// Result of memo validation.
+pub enum VerifyResult {
+    /// Memo has changed and needs to be recomputed.
+    Changed,
+
+    /// Memo remains valid.
+    ///
+    /// Database keys in the hashset represent cycle heads encountered in validation; don't mark
+    /// memos verified until we've iterated the full cycle to ensure no inputs changed.
+    Unchanged(FxHashSet<DatabaseKeyIndex>),
+}
```

Comment:
This change (`bool` being replaced by `VerifyResult`) also seems like it will make non cycle queries pay a hefty perf price. That type has a destructor and is 40 bytes where as the previous return value could fit into a register, this can't.

Reply:
I'll focus first on getting consistently right behavior (including across threads) and then revisit this and other possible ways to claw back perf. Off the top of my head I'm not sure how to improve this, will need to think about it.

Reply:
Yes of course, I am mainly noting down what I saw, no need to tackle perf nits/reviews until things are working properly.